### PR TITLE
opencv: add android support and additional options

### PR DESCRIPTION
**Pull request by andioz**
Specify library name and version:  **opencv/4.x**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

This is my first attempt for a pull request, please bear with me. Regarding issue https://github.com/conan-io/conan-center-index/issues/3715, I try to extend the recipe for Android. 

Known issues:
* I have to disable `openexr`, because this would override the `c++_static` setting in favor of using the shared library instead.
* I have to disable Nvidia support (WITH_CAROTENE = FALSE), because I cannot link to `tegra_hal.a` with cmake target based projects
* Conan hooks reports some errors (red messages), I think they are related to the `android_ndk_installer` used in the Android profile
* I added a `android_profile` to test_package directory, don't know if this is a good idea

I hope I can get help for cleaning up this PR and maybe for enabling the WITH_CAROTENE feature, which would give a performance boost on Android devices (I hope so).

Credits: I used this setting in my local recipe for an older OpenCV version, and I think most of the ideas are from https://github.com/conan-community/conan-opencv.

### CODE REVIEW

**Comment by andioz** [comment]
I added some additional setting, most of it for Android, but some for general purpose, which should be reviewed carefully. I tried solve the failing builds by setting OPENCV_PYTHON_SKIP_DETECTION = True, but there are still failing builds with warnings "CMake Warning: Manually-specified variables were not used by the project: CMAKE_EXPORT_NO_PACKAGE_REGISTRY". I wonder, how to solve this? I'm pretty sure this is not introduced by my changes? Well I hope so, cannot see a reason for that. 

@ericLemanissier, how did you manage to have no failing builds in your last PR?

**Review by ericLemanissier — COMMENTED** [review]
It looks like the error happens during execution of test program on windows. This could be a consequence of setting the default parallel option value to pthread, which does probably not mix well with msvc. You should probably not change this default, and raise an exception in build(self) if the option value is not supported by the OS.

**Comment by andioz** [comment]
Thanks for the hint, I will change the default to False as before in my next commit. But I'm sure I had the errors before I introduced pthreads. And as far I can say from reading the CMakeLists.txt, this option is disabled internally for platforms not supporting it. 

And anyway, I don't understand the warning "CMAKE_EXPORT_NO_PACKAGE_REGISTRY" in general :confused:

File: recipes/opencv/4.x/conanfile.py
Comment (prince-chrismc) [inline recipes/opencv/4.x/conanfile.py:266]: This was not required previously 🤔

File: recipes/opencv/4.x/test_package/android_profile
Comment (prince-chrismc) [inline recipes/opencv/4.x/test_package/android_profile:1]: I believe this file can be remove

Comment (SSE4) [reply recipes/opencv/4.x/test_package/android_profile:1]: agree, not needed

File: recipes/opencv/4.x/conanfile.py
Comment (andioz) [reply recipes/opencv/4.x/conanfile.py:266]: Right, I will remove it. I thought the messages generated without this are the reasons for many failing fests, but I guess it is not.

File: recipes/opencv/4.x/test_package/android_profile
Comment (andioz) [reply recipes/opencv/4.x/test_package/android_profile:1]: Yeah, I'll remove it. I was not sure how to deal with files used for builds other than LInux/Windows/MacOS. I guess we don't keep them at the moment?

Comment (prince-chrismc) [reply recipes/opencv/4.x/test_package/android_profile]: From what I can tell, they are built into C3i system such that every recipe uses the same profiles

File: recipes/opencv/4.x/conanfile.py
Comment (prince-chrismc) [inline recipes/opencv/4.x/conanfile.py:74]: I assume `parallel` with tbb or openmp is invalid on android? perhaps there should be an exception to let consumers know?

Comment (andioz) [reply recipes/opencv/4.x/conanfile.py:74]: As far as I understand OpenCV's CMakeLists.txt, they have several options for parallel execution, filtered internally by several conditions like platform. TBB has a filter for iOS and WinRT (VISIBLE_IF NOT IOS AND NOT WINRT), OpenMP has no filter defined (VISIBLE_IF TRUE), therefore I assume it is legal for Android? pthreads is available for all non-Windows platforms(VISIBLE_IF NOT WIN32 OR MINGW).

Finally, all options are prioritized and the "best" of the enabled and available is selected:

```cmake
ocv_build_features_string(parallel_status EXCLUSIVE
  IF HAVE_TBB THEN "TBB (ver ${TBB_VERSION_MAJOR}.${TBB_VERSION_MINOR} interface ${TBB_INTERFACE_VERSION})"
  IF HAVE_HPX THEN "HPX"
  IF HAVE_OPENMP THEN "OpenMP"
  IF HAVE_GCD THEN "GCD"
  IF WINRT OR HAVE_CONCURRENCY THEN "Concurrency"
  IF HAVE_PTHREADS_PF THEN "pthreads"
  ELSE "none")
status("")
status("  Parallel framework:" "${parallel_status}")
```
 Therefore, I would assume an exception rule would not be easy to define, the cmake file disables unsupported options silently and shows in the summary the used one.

BTW, WITH_PTHREADS_PF is ON by default as most general fallback, I think it would be good to do the same in the recipe?

File: recipes/opencv/4.x/conanfile.py
Comment (martin-schulze-vireso) [inline recipes/opencv/4.x/conanfile.py:310]: For some reason the OpenCV build on my system omits the architecture folder and uses `staticlibs` directly. Can we have both?

File: recipes/opencv/4.x/conanfile.py
Comment (martin-schulze-vireso) [inline recipes/opencv/4.x/conanfile.py:313]: I had to add cpufeatures as a dependency for opencv_core to get it in the correct link order. Maybe this is the problem here too?

**Review by martin-schulze-vireso — COMMENTED** [review]
My own local trials lead to similar changes. I have a few extra observations which I would like to work into this PR, if possible.

Just one question: Which environment did you use for development?

I have a CMake based project (with the cmake conan wrapper) which uses OpenCV and is pulled in as an external native build by an android gradle project. I added `android_ndk_installer/r21d@bincrafters/stable` as a dependency  to get the build going and had to map `CMAKE_SYSTEM_PROCESSOR` manually for everything except x86_64 to get the correct architecture while building.

File: recipes/opencv/4.x/conanfile.py
Comment (prince-chrismc) [reply recipes/opencv/4.x/conanfile.py:74]: beautiful answer 🌪️ , if opencv does a best effort, I think the correct implementation is perfectly adequate. from a consumer perspective if I told opencv through Conan or CMake directly this will have very similar behaviors which matches my preference!

Comment (andioz) [reply recipes/opencv/4.x/conanfile.py:74]: So, this is ok for you? You marked lines 73-74 (openexr exclusion), any question about this?

File: recipes/opencv/4.x/conanfile.py
Comment (andioz) [reply recipes/opencv/4.x/conanfile.py:310]: I'm on Ubuntu 20.04, and we use `android_ndk_installer/r20@bincrafters/stable`, one revision older than yours. I have a pure cmake/make project, which uses my own helper functions to install/update conan packages defined in a project specific `conanfile.py`. And we use a profile file for android:

```ini
include(default)

[settings]
# TODO multiple archs? armv7, armv7s, arm64, x86_64, etc?
arch=armv8
compiler=clang
compiler.libcxx=libc++
compiler.version=8
os=Android
os.api_level=21

[build_requires]
android_ndk_installer/r20@bincrafters/stable

[options]
[env]
```

Thats all, no cmake toolchain file as far as I know. I have no idea what's the issue with `CMAKE_SYSTEM_PROCESSOR`? 

> For some reason the OpenCV build on my system omits the architecture folder and uses staticlibs directly. Can we have both?

Hm, I think before adding extra include directories we should try to understand why this happens? Maybe some flaws in the configuration we should fix instead?

In general, I think we should do this in a separate step, otherwise we get confused. What do you think?

File: recipes/opencv/4.x/conanfile.py
Comment (andioz) [reply recipes/opencv/4.x/conanfile.py:313]: Currently I have no idea how to add `cpufeatures.a` (and `tegra_hal.a` when I enable CAROTINE) in a `conan style` way, yes I think there is a problem with link order, we need some kind of internal dependency to a non-target library. would appreciate any help on this!

File: recipes/opencv/4.x/conanfile.py
Comment (prince-chrismc) [reply recipes/opencv/4.x/conanfile.py:74]: Nope! Feel free to make this resolved

File: recipes/opencv/4.x/test_package/android_profile
Comment (andioz) [reply recipes/opencv/4.x/test_package/android_profile]: But none for Android and iOS of course, right?

File: recipes/opencv/4.x/conanfile.py
Comment (martin-schulze-vireso) [reply recipes/opencv/4.x/conanfile.py:313]: I simply added cpufeatures as a dependency to core and added `self.cpp_infos.components["cpufeatures"].libs=["cpufeatures"]`.

File: recipes/opencv/4.x/conanfile.py
Comment (martin-schulze-vireso) [reply recipes/opencv/4.x/conanfile.py:310]: I already saw this behavior in the bincrafters recipe. IIRC the problem is related to Multiarchitecture ndk vs single architecture but it has been a while.  I believe it was triggered by some cmake variable beeing empty in opencv.

File: recipes/opencv/4.x/conanfile.py
Comment (andioz) [reply recipes/opencv/4.x/conanfile.py:313]: Hm, I think I tried something similar and failed, but not sure. Have to re-check. I wonder, why do I not get a linker error for missing `cpufeatures` at all right now? I don't include it, maybe it is part of the NDK already? Anyway, I have to dig into this again.

File: recipes/opencv/4.x/test_package/android_profile
Comment (prince-chrismc) [reply recipes/opencv/4.x/test_package/android_profile]: macOS is supported but no not android

File: recipes/opencv/4.x/conanfile.py
Comment (prince-chrismc) [reply recipes/opencv/4.x/conanfile.py:313]: You would also need the "main" component to require the `cpufeatures`

File: recipes/opencv/4.x/conanfile.py
Comment (andioz) [reply recipes/opencv/4.x/conanfile.py:310]: would be nice if you can remember which variables influences this, than we can define the path correctly 😏

File: recipes/opencv/4.x/conanfile.py
Comment (martin-schulze-vireso) [reply recipes/opencv/4.x/conanfile.py:313]: IIRC I only got linker errors for cpufeatures in static builds.

File: recipes/opencv/4.x/conanfile.py
Comment (martin-schulze-vireso) [reply recipes/opencv/4.x/conanfile.py:310]: Sry, was on my phone. I looked into OpenCV's cmake now and I'd say the problem is in [`cmake/OpenCVInstallLayout.cmake`](https://github.com/opencv/opencv/blob/4.5.0/cmake/OpenCVInstallLayout.cmake#L10) which defines:

```
ocv_update(OPENCV_LIB_ARCHIVE_INSTALL_PATH    "sdk/native/staticlibs/${ANDROID_NDK_ABI_NAME}")
```

Which in turn is used in `cmake/OpenCVModule.cmake` for `macro(_ocv_create_module)`, setting the `ARCHIVE DESTINATION`:

```
ocv_install_target(${the_module} EXPORT OpenCVModules OPTIONAL
      RUNTIME DESTINATION ${OPENCV_BIN_INSTALL_PATH} COMPONENT libs
      LIBRARY DESTINATION ${OPENCV_LIB_INSTALL_PATH} COMPONENT libs NAMELINK_SKIP
      ARCHIVE DESTINATION ${OPENCV_LIB_ARCHIVE_INSTALL_PATH} COMPONENT dev
      )
```

`ANDROID_NDK_ABI_NAME` is defined in `platforms/android/android.toolchain.cmake` (just a copy of the default android toolchain file), which I presume not to be loaded in my case. I don't see it being included by OpenCV's CMake scripts.

Comment (andioz) [reply recipes/opencv/4.x/conanfile.py:310]: I did some "prinf debugging" (well, ok, more "message" debugging inside cmake files). I can see that at the beginning of OpenCV's main CMakeLists.txt `ANDROID_NDK_ABI_NAME` is set to `arm64-v8a` with my android profile in use. Furthermore, I found out that it doesn't come from OpenCV's `android.toolchain.cmake`, which seems not to be used at all. But instead it comes from the same file in the `android_ndk_installer`! (r20 in my case).

SO, AFAIK it looks like this variable should always be set, otherwise it looks like some error to me. The update of the `OPENCV_LIB_ARCHIVE_INSTALL_PATH` as you show above seems to assume an always set value for `ANDROID_NDK_ABI_NAME` I would say.

Comment (martin-schulze-vireso) [reply recipes/opencv/4.x/conanfile.py:312]: I looked into that a bit more and noticed that the problem only occured when using the cmake wrapper instead of the profile. The problem was that I did not specify the NDK as build_requires. I assume conan then just picked up the system NDK and did somehow ignore their toolchain file.

> I can see that at the beginning of OpenCV's main CMakeLists.txt ANDROID_NDK_ABI_NAME is set to arm64-v8a

There is nothing in OpenCV CMakeLists.txt that does this. The most likely behavior is that CMake itself loads the android toolchain file (probably instructed to do so by conan).

Comment (andioz) [reply recipes/opencv/4.x/conanfile.py:312]: > There is nothing in OpenCV CMakeLists.txt that does this. The most likely behavior is that CMake itself loads the android toolchain file (probably instructed to do so by conan).

Well, using the the android profile specifying an android_ndk_installer is one way to get this variable from this wrapper. This is the way my project works. If you are using CMakeLists.txt directly, I guess you have to specify the toolchain file manually at the command line. I remeber we had to use this years ago for some embedded /cross compile projects too.

I think I mark this as resolved, don't know if there is some action required for this, ok?

File: recipes/opencv/4.x/conanfile.py
Comment (andioz) [reply recipes/opencv/4.x/conanfile.py:317]: Thanks for the hint, I added `cpufeatures` now to `opencv_core`, where it is used. Now the link order looks good.

File: recipes/opencv/4.x/conanfile.py
Comment (martin-schulze-vireso) [reply recipes/opencv/4.x/conanfile.py:312]: Okay from my end.

**Review by prince-chrismc — APPROVED** [review]
LGTM! Not an android expert so well see what other reviewers say 🤞

File: recipes/opencv/4.x/conanfile.py
Comment (SSE4) [inline recipes/opencv/4.x/conanfile.py:305]: you can check `self.settings.os.api_level`

File: recipes/opencv/4.x/conanfile.py
Comment (SSE4) [inline recipes/opencv/4.x/conanfile.py:312]: you can check `self.settings.arch` and `self.settings.os`

**Comment by martin-schulze-vireso** [comment]
I tried this PR for static builds and it worked on following devices (others were not tested):

- x86_64 Linux (clang)
- arm64 Android

File: recipes/opencv/4.x/conanfile.py
Comment (andioz) [reply recipes/opencv/4.x/conanfile.py:305]: Ok, I'll try to use the internal rules for this using `self.settings.os.api_level`, this looks simple enough.

File: recipes/opencv/4.x/conanfile.py
Comment (andioz) [reply recipes/opencv/4.x/conanfile.py:312]: Hm, this one is more complicated I guess, and I'm afraid this is a little bit too much for this first step, supporting Android. the rule inside OpenCV's CMakeLists.txt is:

```cmake
OCV_OPTION(WITH_CAROTENE "Use NVidia carotene acceleration library for ARM platform" ON
  VISIBLE_IF (ARM OR AARCH64) AND NOT IOS AND NOT (CMAKE_VERSION VERSION_LESS "2.8.11"))
```

There is a condition for ARM, AARCH64, IOS and CMake Version... Trying to model this in the recipe looks like more code than just one line or two. A more simpler approach would be: can we check if a library `tegra_hal.a` exists in the `lib` folder and add it if it is available? Then we could avoid to rebuild the internal rule (which may change in future versions).

At the bottom line, if we cannot find a simple way to add this conditionally, I would to prefer to disable it and let it open for later improvements, when more other platforms will be added. 

What do you think?

**Comment by andioz** [comment]
This one failed, but I cannot see any information why or on which platform? Does this mean I have a general error in the recipe?

File: recipes/opencv/4.x/conanfile.py
Comment (andioz) [reply recipes/opencv/4.x/conanfile.py:312]: I think I found the solution, using tools.collect_libs(). This removes all the rule reproduction, is shorter and clean.

Comment (SSE4) [reply recipes/opencv/4.x/conanfile.py:312]: still, conditional might be needed for the options above?
but if it's too complicated, it's okay to omit for the first iteration, so up to you

Comment (andioz) [reply recipes/opencv/4.x/conanfile.py:312]: I'm just waiting for the current check to pass, then I push my next version. For `libmediandk.so`, the condition is still required, yes, but this is simple enough (only NDK API level dependency), and others go away completely. I hope this will be the best solution then.

**Comment by andioz** [comment]
Looks like I have to be up-to-date with upstream. After merging upstream master into master and then master into branch the checks are running.

> I successfully tried this PR for shared=False on following devices (including running an automatic test suite and some manual tests on the Android device):
> 
>     * x86_64 Linux (clang)
> 
>     * arm64 Android 10 (NDK r21d)
> 
>     * x86 and x86_64 Android 10 emulator (NDK r21d)

Many thanks for doing the extra testing, I don't have all these platforms available.

**Comment by andioz** [comment]
@SSE4, @prince-chrismc do you have an idea why the last checks fails after a minute or so, without any information and email notification? I reverted no to my yesterdays version, which passed before. No idea how to continue....

**Comment by prince-chrismc** [comment]
C3i has been struggling the last two days on big PRs, I am sure the infrastructure team is flooded with error message on their side.

I've had a few PR's fail similarly, shouldn't be long before the CCI team sorts out the issue.

You can always manually trigger by closing, waiting 10 seconds and re-opening the PR

**Comment by andioz** [comment]
Thanks, good to know. I'll wait some time and retry from time to time.

**Review by prince-chrismc — APPROVED** [review]
LGTM! 🚀 very nice addition

**Review by SSE4 — APPROVED** [review]

**Comment by andioz** [comment]
Hi there, just to be sure, us there anything I should do now? Or just wait for approval? 😏

**Comment by prince-chrismc** [comment]
Just wait 🕥 ! There was a ton of small PRs opened so it hides this one but I'm sure that one missing review will come soon

**Comment by SSE4** [comment]
> Hi there, just to be sure, us there anything I should do now? Or just wait for approval? 😏

yes, just wait for the approval from @uilianries and someone else. it may take time since there are more than 100 PRs opened right now.

**Review by uilianries — APPROVED** [review]

**Comment by andioz** [comment]
@prince-chrismc one last question (I hope): after merging an error occurred (continuous-integration/jenkins/branch)  and the recipe in conan center is not updated yet. Another temporal problem in the build system? Will it be updated automatically? Something I can do?

**Comment by prince-chrismc** [comment]
Unfortunately yes. again projects at the size of OpenCV sometimes edge out the scaling/use cases. There is a nightly job which tries  fix missing packages. 

I have not seen any information on it, but they are usually good at fixing them. @ jgsogo and @ danimtb I believe are on the infrastructure team and are the one's I usually ping for this kinda problem

**Comment by andioz** [comment]
> Unfortunately yes. again projects at the size of OpenCV sometimes edge out the scaling/use cases. There is a nightly job which tries fix missing packages.
> 
> I have not seen any information on it, but they are usually good at fixing them. @ jgsogo and @ danimtb I believe are on the infrastructure team and are the one's I usually ping for this kinda problem

Thanks for the information. So, I'll wait again :smile:. Let me know if I can help in any way.

**Comment by martin-schulze-vireso** [comment]
The recipe on conan center is still not updated. 

@prince-chrismc Did you mean for @andioz to ping @jgsogo and @danimtb or are they already working on this behind the scenes?

**Comment by andioz** [comment]
> The recipe on conan center is still not updated.
> 
> @prince-chrismc Did you mean for @andioz to ping @jgsogo and @danimtb or are they already working on this behind the scenes?

Same idea at the same time - i added an issue https://github.com/conan-io/conan-center-index/issues/3879

**Comment by jgsogo** [comment]
Ready to serve!

image

**Comment by prince-chrismc** [comment]
He would need to ping them, creating an issue works well too!

I am just giving orders! 
image

**Pull request merged**
